### ISSUE TITLE
Local Gradient Accumulation is slower than the PyTorch implementation.

### ISSUE THREAD

**Issue by cirquit** [learning-at-home/hivemind#566]
I think I found a slight performance issue with Hivemind. A call to `opt.step()` before the TBS is reached that accumulates the gradients is somehow not as performant as the native PyTorch gradient accumulation. There is a trend visible based on the model parameter count, so I presume it's not DHT related.

Here's the experimental "proof". The single GPU experiment is a baseline without hivemind, the 2,3,4, and 8 GPU runs are with hivemind. The first figure shows the average `backward_s` timing, where no real change in timings can be seen (this is the call in the 1 GPU experiment that does the gradient accumulation; the accumulation for the 2-8 GPU runs happens over `opt.step()`). 
This means that accumulating the gradients is more or less 0 cost for the baseline PyTorch implementation.

![image](https://user-images.githubusercontent.com/9363632/235985770-4b6af0b6-aee9-4bd0-8bf6-b509bad1a7d2.png)

The second figure shows the no-sync `opt.step()` call timings for the 2-8 GPU runs, which show a trend of increased slowdown the bigger the model gets, suggesting something depending on the model parameter count. Maybe there's a GPU->CPU memory copy or GPU internal copy happening?

![image](https://user-images.githubusercontent.com/9363632/235985635-644c3b32-896e-4b7c-a63f-fa35e25795eb.png)

I also compared the actual throughput impact of this slower no-sync `opt.step()` , and it reaches at worst only 48% (ConvNextLarge) and at best 78% (ResNet152) of the baseline performance. The throughput compared was the normalized local hivemind throughput (the one without averaging) and the baseline 1-GPU throughput.

![image](https://user-images.githubusercontent.com/9363632/235986816-12964ccf-5bdc-4912-8012-f81202c30227.png)

I think the relevant code is here https://github.com/learning-at-home/hivemind/blob/master/hivemind/optim/grad_averager.py#L129-L148
